clang/include/clang/Sema/Sema.h

```diff
@@ -12971,7 +12971,15 @@ void Sema::addIntelSYCLSingleArgFunctionAttr(Decl *D,
         return;
       }
     }
-    if (CI.getParsedKind() == ParsedAttr::AT_SYCLIntelMaxGlobalWorkDim) {
+    if (CI.getParsedKind() == ParsedAttr::AT_SYCLIntelNoGlobalWorkOffset) {
+      if (ArgInt > 1) {
+        Diag(E->getExprLoc(),
```

Comment:
Sorry for not noticing this earlier, but I was curious about the warning since it wasn't familiar to me and that caused me to notice a few issues.

https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/DiagnosticSemaKinds.td#L11147

0. The diagnostic suggests that the boolean argument here will be adjusted to be `1` automatically, but the code is dropping the attribute entirely.

1. The diagnostic wording doesn't follow the usual style (it starts with a capital letter and contains sensible full stop punctuation). This isn't specific to your patch, but is an inconsistency we may want to do something about (I notice a few other warnings that have similar issues).

2. The warning itself is rather suspect. Generally speaking, a boolean value is either zero or nonzero, so warning "you wrote something nonzero, we've switched it to a different nonzero value" is a bit strange. This also isn't specific to your patch.

Reply:
>>Generally speaking, a boolean value is either zero or nonzero, so warning "you wrote something nonzero, we've switched it to a different nonzero value" is a bit strange. 

Yes, I agree. This diagnostic seems very confusing to me.  This diagnostic was added only for this attribute.

I am inclined to change like this. @AaronBallman what is your thought? This provides clear info that the attribute can only take 0 or one.

if (ArgInt > 1) {
        Diag(E->getBeginLoc(), diag::err_attribute_argument_out_of_range)
            << CI.getAttrName() << 0 << 1 << E->getSourceRange();
        return;
      }

Reply:
Backing up a step so I can see the bigger picture -- what is the goal of diagnosing a value other than 0 or 1 here? Is that a requirement from some other standard? Do bad things happen if they pass 2 instead of 1? Or is it more a matter of trying to enforce a sane coding style?

Reply:
I am not sure about this. Tagging @MrSidims for this comments. He reviewed original attribute implementation.

Reply:
I would imagine that if any nonzero value will do (so it's just a matter of style), we should accept silently and not bother warning. If the value has to be zero or one because of assumptions elsewhere in the compiler, we could force the value 0 or 1 when creating the semantic attribute (so if the user passes in `2`, we manually create an integer literal for `1` instead). If the value has to be zero or one because of portability to other toolchains, requirements from another standard, etc then diagnosing makes more sense to me.

Reply:
Tagging @MrSidims/@GarveyJoe for this comment.

Reply:
> Backing up a step so I can see the bigger picture -- what is the goal of diagnosing a value other than 0 or 1 here? Is that a requirement from some other standard? Do bad things happen if they pass 2 instead of 1? Or is it more a matter of trying to enforce a sane coding style?

I think you pointed the very problem. The API is unclear. If a boolean is expected, then it should be `[[intelfpga::no_global_work_offset(true)]]` and `[[intelfpga::no_global_work_offset(false)]]`, which removes the need for testing anything less just by relying on usual C/C++ conventions.

Reply:
I don't think this diagnostic carries its weight. It introduces an inconsistency with how other attributes are handled (which makes all attributes harder to use because you have to guess which attributes work what way) and doesn't follow the usual C or C++ conventions (where 0 is false and nonzero is true). Also, while the SYCL attributes currently focus on C++, part of the consistency with C means that `true` and `false` are actually problematic in some C language modes because those aren't keywords (you need to include a header file to use macros with that name, and the macros expand to integers anyway).

I'd recommend removing this diagnostic entirely.

Reply:
@AaronBallman for the recommendation.  I will remove this entirely. 

I have one question, since attribute takes only 0 and 1, if user passes 2 or more as attribute value, should we give the diagnostic like and ignore the attribute. We use this for other attributes also.

if (ArgInt > 1) {
Diag(E->getBeginLoc(), diag::err_attribute_argument_out_of_range)
<< CI.getAttrName() << 0 << 1 << E->getSourceRange();
return;
}

Reply:
I wouldn't diagnose that situation at all because the attribute takes a Boolean expression, and `2` is a valid Boolean expression that means `true`. If it makes your life easier, you could correct the value when creating the semantic attribute. e.g.,
```
Expr *Arg = nullptr;
if (Expr *E = Attr.getArgAsExpr(0)) {
  uint32_t Val;
  if (!checkUint32Argument(S, AL, E, Val))
    return;
  // The optional argument is present but does not evaluate to 0 or 1,
  // so force the value to be 1.
  if (Val && Val != 1)
    Arg = IntegerLiteral::Create(...); // Create literal 1
  else
    Arg = E; // The literal the user wrote is fine, just use it.
} else
  Arg = IntegerLiteral::Create(...); // Create literal 1 for the default value

D->addAttr(..., Arg, ...);
```
Note, I'd only jump through these hoops if you absolutely have to as it makes the code more complicated to reason about. A better approach, if possible, is to handle it when getting the value out of the expression attached to the semantic attribute. At some point you'll be getting an `APInt` object out of the expression and you can check whether the value of that integer is nonzero or not at that point.

Reply:
I agree. This will make the code more complicated. I have removed the diagnostic entirely.  If there is a reason, we can make the above changes and could correct the value when creating the semantic attribute. 

I see similar HLS attribute that supports GNU syntax and uses old spelling of SYCL attribute (uses_global_work_offset, https://github.com/intel/llvm/pull/1090). This attributes treats the value 2 or so  as "true" Boolean value and generates same AST/IR like value 1.